### PR TITLE
Allow exclude extras from user agent

### PR DESCRIPTION
**Pull request by vz10**
Add ability to exclude extras from UserAgent

### CODE REVIEW

**Review by kyleknap — COMMENTED** [review]
Looks fine. Just had one comment about the interface/user agent value we produce. Other than that let's get the IMDS user agent updated for the credential provider and make sure to have a test for that.

File: botocore/session.py [hunk 1]
```diff
@@ -457,7 +457,7 @@ def user_agent(self):
                                           platform.release())
         if os.environ.get('AWS_EXECUTION_ENV') is not None:
             base += ' exec-env/%s' % os.environ.get('AWS_EXECUTION_ENV')
-        if self.user_agent_extra:
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply botocore/session.py:460]: Yea I agree lets go as minimal as possible.

**Comment by stealthycoin** [comment]
Will need to update here: https://github.com/boto/botocore/blob/v2/botocore/credentials.py#L83

File: tests/unit/test_credentials.py [hunk 2]
```diff
@@ -1716,6 +1718,17 @@ def test_custom_cache(self):
         cache = resolver.get_provider('assume-role').cache
         self.assertIs(cache, custom_cache)
 
+    @mock.patch('botocore.awsrequest.AWSRequest')
+    def test_imds_use_truncated_user_agent(self, request):
+        session = Session()
```
Comment (stealthycoin) [inline tests/unit/test_credentials.py:1723]: Does it not work to use the pattern the prior tests use where we pull out a provider using 

```python
        resolver = credentials.create_credential_resolver(self.session)
        provider = resolver.get_provider('iam-role')
        provider.load()
```

I would imagine we shouldn't have to mock out the awsrequest and instead be able to inject mocks.

**Review by stealthycoin — COMMENTED** [review]
The implementation looks good, I just had a comment about the tests.

**Review by stealthycoin — APPROVED** [review]
Looks good.

File: tests/functional/test_credentials.py [hunk 3]
```diff
@@ -828,3 +828,16 @@ def test_credential_process_returns_error(self):
             reg = r"(?s)^((?!b').)*$"
             with self.assertRaisesRegexp(CredentialRetrievalError, reg):
                 session.get_credentials()
+
+
+class TestInstanceMetadataFetcher(unittest.TestCase):
```
Comment (kyleknap) [inline tests/functional/test_credentials.py:833]: Let's subclass from `BaseSessionTest` and use its session if we can. Mainly there are some environment variables that can be set to affect behavior of this provider and we should make sure they do not get inadvertently set.

**Review by kyleknap — COMMENTED** [review]
Looks fine. Just had one more suggestion

**Review by kyleknap — APPROVED** [review]
Thanks! :ship:

**Pull request merged**